zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -229,10 +229,16 @@ where
         .build(consensus_branch_id, &prover)
         .map_err(Error::Builder)?;
 
-    // We only called add_sapling_output() once.
-    let output_index = match tx_metadata.output_index(0) {
-        Some(idx) => idx as i64,
-        None => panic!("Output 0 should exist in the transaction"),
+    let output_index = match to {
+        // Sapling outputs are shuffled, so we need to look up where the output ended up.
+        RecipientAddress::Shielded(_) => match tx_metadata.output_index(0) {
+            Some(idx) => idx as i64,
+            None => panic!("Output 0 should exist in the transaction"),
+        },
+        // This function only spends shielded notes, so there will only ever be a single
+        // transparent output in this case (and even if there were more, we don't shuffle
+        // the transparent outputs).
+        RecipientAddress::Transparent(_) => 0,
```

Comment:
This seems a bit fragile to me - I'm not sure how someone working on the `build` code in the future would know to maintain this invariant.

Reply:
This invariant is maintained by the usage of the builder code here, not its implementation elsewhere; as long as this function isn't spending transparent coins, it's fine (and someone making such a change should see this very visible comment). But if there _were_ changes to the builder implementation that caused more transparent outputs to be created than were requested (IDK why we would do that), it would be a material change that should be documented in the changelog, and thus something the downstream consumer of the builder implementation (i.e. us in this case) should notice.

Reply:
I thought about this again, and decided it was easy enough to detect this properly by scanning the built transaction.